packages/appium/lib/appium.js

```diff
@@ -410,6 +675,19 @@ class AppiumDriver extends DriverCore {
         );
         await driverInstance.updateSettings(jwpSettings);
       }
+
+      // if the user has asked for bidi support, and the driver says that it supports bidi,
+      // send our bidi url back to the user. The inner driver will need to have already saved any
+      // internal bidi urls it might want to proxy to, cause we are going to overwrite that
+      // information here!
+      if (dCaps.webSocketUrl && driverInstance.doesSupportBidi) {
+        const {address, port, basePath} = this.args;
+        const bidiUrl = `ws://${address}:${port}${basePath}${BIDI_BASE_PATH}/${innerSessionId}`;
+        // @ts-ignore webSocketUrl gets sent by the client as a boolean, but then it is supposed
+        // to come back from the server as a string. TODO figure out how to express this in our
+        // capability constraint system
```

Comment:
we can mark it as `boolean|string|null`

Reply:
unfortunately the type of `dCaps.webSocketUrl` is set from the constraints, which only allows one type, not a disjunctive type, as far as i can tell.